kratos/utilities/volume_inside_voxel_utility.cpp

```diff
@@ -0,0 +1,382 @@
+            } 
+        }
+        return volume;
+    }
+
+    /***********************************************************************************
+     **********************************************************************************/
+
+    double VolumeInsideVoxelUtility::EdgesApproximation(const GeometryType& rVoxel) {
+        double volume = 0;
+        GeometryArrayType Edges = rVoxel.GenerateEdges();
+        PointsArrayType nodes = rVoxel.Points();
+        for (int i = 0; i < Edges.size(); i++) {
+            PointsArrayType ends = Edges[i].Points();
+            if(ends[0].GetSolutionStepValue(DISTANCE) > 0 && ends[1].GetSolutionStepValue(DISTANCE) > 0) {
+                volume+=1.0/Edges.size();
+            } else if(
+                ends[0].GetSolutionStepValue(DISTANCE) > 0 && ends[1].GetSolutionStepValue(DISTANCE) < 0 || 
+                ends[0].GetSolutionStepValue(DISTANCE) < 0 && ends[1].GetSolutionStepValue(DISTANCE) > 0 ) {
+                volume+=1.0/(Edges.size()*2);
+            }
+        }
+        return volume;
+    }
+
+    /***********************************************************************************
+     **********************************************************************************/
+
+    double VolumeInsideVoxelUtility::EdgesPortionApproximation(
+        const GeometryType& rVoxel,  
+        const GeometryArrayType& rTriangles     
+    ) {
+        double volume = 0;
+        GeometryArrayType Edges = rVoxel.GenerateEdges();
+        std::vector<double> Distances;
+
+        for (int i = 0; i < Edges.size(); i++) {
+            Distances.push_back(0);
+            PointsArrayType ends = Edges[i].Points();
+
+            for (auto& r_triangle : rTriangles) {
+                array_1d<double,3> intersection;
+                int result = IntersectionUtilities::ComputeTriangleLineIntersection(r_triangle,ends[0],ends[1],intersection);
+                
+                if(result == 1) {
+                    double Dist = Distance(ends[0], intersection);
+                    Distances.push_back(Dist);
+                }  
+            } 
+            Distances.push_back(Distance(ends[0],ends[1]));
+            std::sort(Distances.begin(),Distances.end());       //WOULD A SET BE MORE EFFICIENT?     
+            double edgePortion = VolumeInsideVoxelUtility::EdgeFilledPortion(Distances, ends);
+            volume += edgePortion/Edges.size();  
+            Distances.clear();              
+        }
+        return volume;
+    }
+
+    /***********************************************************************************
+     **********************************************************************************/
+
+    double VolumeInsideVoxelUtility::VoxelVolume2D(
+        const GeometryType& rFace,  
+        const GeometryArrayType& rTriangles     
+    ) {
+        double Area = 0;
+        GeometryArrayType Edges = rFace.GenerateEdges();
+        PointsArrayType Nodes = rFace.Points(); 
```

Comment:
```suggestion
        const auto& r_nodes = rFace.Points(); 
```
Note the `&` to avoid the copy.